### PR TITLE
Participating in Council Governance is Free for First Time Voters and Successful Closing

### PR DESCRIPTION
**Pull request by shamb0**
Participating in Governance should be Free for First Time Voters

Fixes: https://github.com/paritytech/substrate/issues/6904

**Summary of Code Changes**

| Title | Discp |
| --- | --- |
| Module | pallet-collective |
| Extrinsic functions | vote(), close() |

transaction fee will be waived for extrinsic calls vote() & close() for conditions described below ...

- In **vote()** api, first vote of the member in the motion is free, new logic got introduced to detect the member's first vote ( based on none status of ayes & nays ). and return value DispatchResultWithPostInfo is embedded with pay info ( Pays::No for first vote condition & Pays::No for all other conditions ).

- If the close operation completes successfully (with either approval or disapproval),the transaction fee will be waived. return value DispatchResultWithPostInfo is embedded with Pays::No under successful close conditions.

- Changes got verified with new Unit test function "motions_all_first_vote_free_works()"

### CODE REVIEW

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:1493]: This should panic, forcing the test to fail.

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:1566]: You def dont need this :)

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:1558]: Should be another panic situation, forcing the test to fail

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:553]:
```suggestion
			} else {
```

**Review by shawntabrizi — APPROVED** [review]
Looks good to me!

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:519]: This behavior should be mentioned in the docs above

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:668]: This behavior should be mentioned in the docs above

File: frame/collective/src/lib.rs
Comment (shamb0) [reply frame/collective/src/lib.rs:1492]: Thanks Shawn, replaced with suggested one.

File: frame/collective/src/lib.rs
Comment (shamb0) [reply frame/collective/src/lib.rs:1557]: Thanks Shawn, replaced with suggested one.

File: frame/collective/src/lib.rs
Comment (shamb0) [reply frame/collective/src/lib.rs:519]: Updated the function header Doc string ...

File: frame/collective/src/lib.rs
Comment (shamb0) [reply frame/collective/src/lib.rs:667]: Updated the function header Doc string ...

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:492]:
```suggestion
		/// Requires the sender to be a member.
		///
		/// Transaction fees will be waived if the member is voting on any particular proposal
		/// for the first time and the call is successful.
```

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:566]:
```suggestion
		/// If the close operation completes successfully (with either approval or disapproval),
		/// the transaction fee will be waived.
```

Comment (shawntabrizi) [inline frame/collective/src/lib.rs:566]:
```suggestion
		/// If the close operation completes successfully (with either approval or disapproval),
		/// the transaction fee will be waived.
```

File: frame/collective/src/lib.rs
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:674]:
```suggestion
					Pays::No,
```

Small nit pick detail, we like to put a comma at the end of newline separated lists.

https://users.rust-lang.org/t/trailing-commas/13993

Few more instances of this in other spots in this PR

**Comment by shawntabrizi** [comment]
Looks ready to merge once we have another review and an audit. Audit team is made aware and will look next week.

File: frame/collective/src/lib.rs
Comment (shamb0) [reply frame/collective/src/lib.rs]: Thanks Shawn, will review again and fix it

**Comment by shamb0** [comment]
Fine Shawn, And Trailing Commas are added ...

File: frame/collective/src/lib.rs
Comment (kianenigma) [inline frame/collective/src/lib.rs:1487]:
```suggestion
			// For the motion, acc 2's first vote, expecting Ok with Pays::No.
```

File: frame/collective/src/lib.rs
Comment (kianenigma) [inline frame/collective/src/lib.rs:1500]:
```suggestion
			// Duplicate vote, expecting error with Pays::Yes.
```

File: frame/collective/src/lib.rs
Comment (kianenigma) [inline frame/collective/src/lib.rs:1518]:
```suggestion
			assert_eq!(vote_rval.unwrap_err().post_info.pays_fee, Pays::Yes);
```

File: frame/collective/src/lib.rs
Comment (kianenigma) [inline frame/collective/src/lib.rs:1520]:
```suggestion
			// Modifying vote, expecting ok with Pays::Yes.
```

File: frame/collective/src/lib.rs
Comment (kianenigma) [inline frame/collective/src/lib.rs:1533]:
```suggestion
			// For the motion, acc 3's first vote, expecting Ok with Pays::No.
```

File: frame/collective/src/lib.rs
Comment (kianenigma) [inline frame/collective/src/lib.rs:1546]:
```suggestion
			// acc 3 modify the vote, expecting Ok with Pays::Yes.
```

File: frame/collective/src/lib.rs
Comment (kianenigma) [inline frame/collective/src/lib.rs:1602]:
```suggestion
			assert_eq!(vote_rval.unwrap_err().post_info.pays_fee, Pays::Yes);
```

**Review by kianenigma — APPROVED** [review]
Could use my comments, but all in all looks good.

**Comment by shamb0** [comment]
Thanks @kianenigma for review & comments, will make sure not repeat the minor mistakes again :-)

**Comment by redzsina** [comment]
We are done with our review for this PR - please refer to issue 30 in the internal tracker.

**Review by thiolliere — APPROVED** [review]
I don't have opinion on whether those calls should be free, but code looks good and logic can't be abused

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**
